[EMBR-6365] Add documentation for embrace dashboard backlink for Grafana (#189)
embrace-io/embrace-docs: docs/data-destinations/grafana-cloud-setup.md

- Aggregations. Example: if the metric is aggregated by `os_version` and `country` (i.e. `sum by(app_id, country, os_version, embrace_metric_name) (embrace_crash_hourly_total)`)
Embrace uses those filters with equals operator on the Embrace dashboard.

Embrace doesn't support the following options used on the Grafana visualization:
- The following aggregations are ignored when Embrace generates the dashboard backlink: 
  - `duration_bucket`, `group_id`, `log_property_value`, `moment_property_value`, `root_span_attribute_value`, `root_span_duration_bucket`, 
  `session_property_value`, `status_code`, `tag_value`.
- [Top N](/embrace-api/supported_metrics_and_queries/#dimension-reduction---other) aggregations. Embrace converts these since the `other` value cannot be mapped to a specific value on the Embrace dashboard. Aggregations converted:
  - `device_model` converted to `model_market_name` (Model Name). 
  - `top_n_domain` converted to `domain`.
  - `top_n_market_name` converted to `model_market_name` (Model Name).
  - `top_n_path` converted to `path`.

> - Label filters. Example: if the metric is filtered by `os_version` and `country` (i.e. `sum by(app_id, country, os_version, embrace_metric_name) (embrace_crash_hourly_total{os_version="12", country="US"})`)

Reviewer: Why does this happen? I would think that the labels are still available, but they will always be os_version =12 and coutnry=US.

Author: You are right. Fix it.

Embrace ignores those filters when it generates the dashboard backlink.

> - Filters at creation time, when you create the custom metric on the Embrace dashboard, you can specify filters, Embrace 

Reviewer: ```suggestion
- Filters at creation time. When you create the custom metric on the Embrace dashboard, you can specify filters, Embrace 
```

> don't use those filters when it generates the dashboard backlink. Example: you may create a custom metric `sessions_total` group by `os_version` and filter by `app_version = 1.2.3` on 

Reviewer: ```suggestion
doesn't use those filters when it generates the dashboard backlink. Example: you may create a custom metric `sessions_total` group by `os_version` and filter by `app_version = 1.2.3` on 
```

<img src={require('@site/static/images/data-destinations/grafana_cloud_prometheus_data_source.png').default} alt="Image showing Grafana Cloud datasource" />

<img src={require('@site/static/images/data-destinations/grafana_cloud_prometheus_data_source_selection.png').default} alt="Image showing the Prometheus data source selection" />

### Embrace Metrics API vs. Data Destination

- **Embrace Metrics API**: Alternatively, you can configure Grafana to pull metrics from Embrace using PromQL queries via the Embrace Metrics API. However, this method requires Grafana to be set up with the Embrace Metrics API as a data source and is typically only used if you have specific requirements for on-demand metrics retrieval.

- **Data Destination**: Embrace can push metrics directly to your Grafana Cloud instance as a data destination. This setup leverages Grafana Cloud's managed Prometheus services, enabling seamless integration without the need for an additional Prometheus instance. This method is ideal for most users who prefer a simple, automated approach to access Embrace metrics within Grafana Cloud.

## Embrace dashboard backlink

> Embrace provides this feature to allows the customer to navigate from Grafana visualizations using Embrace metrics to
> the Embrace dashboard.

Reviewer: ```suggestion
Embrace provides a backlinking feature that allows the customer to navigate from Grafana visualizations using Embrace metrics to
the Embrace dashboard.
```

> Embrace exposes a public endpoint that based on the time range and aggregations used on the Grafana visualizations will 
> redirect the customer to the Embrace dashboard with the same time range and aggregations. Embrace redirects the customer
> to different pages based on the metric that they are visualizing. If the metric is a sessions metric, Embrace redirects the 
> customer to the 

Reviewer: ```suggestion
Embrace exposes a public endpoint that accepts the time range and aggregations used on the Grafana visualizations and will
redirect the customer to the Embrace dashboard with those same time range and aggregations. Embrace redirects the customer
to different pages based on the metric that they are visualizing. If the metric is a sessions metric, Embrace redirects the 
customer to the 
```

> customer to the 

Reviewer: to the?

Author: done

### What does Embrace support and not support?

Embrace supports the following options used on the Grafana visualization:
- Time range. Example: if the time range on Grafana visualization is last 12 hours (i.e. now is `2025-02-04 08:45:09`),
Embrace uses the same time range on the Embrace dashboard (i.e. start date is `2025-02-03 23:00` and end date is `2025-02-04 12:00`).

> - Aggregations. Example: if the metric is aggregated by `os_version` and `country` (i.e. `sum by(app_id, country, os_version, embrace_metric_name) (embrace_crash_hourly_total)`)

Reviewer: I think the Grafana uses the term "labels" instead of aggregations. When you were setting up the dashboard in Grafana which word did it use?

Author: Aggregations.

Embrace uses those filters with equals operator on the Embrace dashboard.

Embrace doesn't support the following options used on the Grafana visualization:

> - Some aggregations are not supported on the Embrace dashboard as filters, Embrace skips them on the Embrace dashboard backlink. Aggregations not supported: 

Reviewer: - The following labels are ignored when generating the backlink.

Author: done

>   - `duration_bucket`, `group_id`, `log_property_value`, `moment_property_value`, `root_span_duration_bucket`, `root_span_attribute_value`, 

Reviewer: nit: `root_span_attribute_value` should go before `root_span_duration_bucket`

Author: done

  `session_property_value`, `status_code`, `tag_value`.

> - Some aggregations are converted to other filters on the Embrace dashboard. The next 4 filters have the [top_n](/embrace-api/supported_metrics_and_queries/#dimension-reduction---other) feature, 

Reviewer: - Top N aggregations. We convert these since the `other` value cannot be mapped to a specific value on the Embrace dashboard. Aggregations converted:

Author: done

if the value is `other`, there is no way to map it to a specific value on the Embrace dashboard and Embrace skips them. Aggregations converted: 
  - `device_model` converted to `model_market_name` (Model Name). 
  - `top_n_domain` converted to `domain`.
  - `top_n_market_name` converted to `model_market_name` (Model Name).
  - `top_n_path` converted to `path`.

> - Label filters. Example: if the metric is filtered by `os_version` and `country` (i.e. `sum by(app_id, country, os_version, embrace_metric_name) (embrace_crash_hourly_total{os_version="12", country="US"})`)

Reviewer: Not sure what you mean here

Embrace doesn't use those filters on the Embrace dashboard.

> - Filters at creation time, when you create the custom metric on the Embrace dashboard, you can specify filters, Embrace 

Reviewer: - Filters not present in the labels. Example: You may create a custom metric `sessions_total` group by `os_version` and filter by `app_version = 1.2.3`. The metric will only provide the `os_version` label, and the `app_version` label will not be present.

don't use those filters on the Embrace dashboard. Example: if you created a custom metric `sessions_total` group by `os_version` and filter by `app_version = 1.2.3` on 
the Embrace dashboard, and you graph it on Grafana as `sum by (app_id, os_version, embrace_metric_name) (embrace_sessions_hourly_total)`,
Embrace uses the `os_version` filter but not the `app_version` filter on the Embrace dashboard.

### How to configure it?

1. Go to the Grafana Dashboard where you visualize the Embrace metrics.
2. Go to the Grafana Visualization where you want to add the Embrace dashboard backlink. 
   1. All the metrics must be aggregated by "app_id" and "embrace_metric_name". Embrace uses those aggregations to generate 
      the Embrace dashboard backlink. Example: this (`sum by (os_version) (embrace_sessions_hourly_total)`) doesn't work and this 
     (`sum by (os_version, app_id, embrace_metric_name) (embrace_sessions_hourly_total)`) works.
3. Click the "three dots" at the top right of the Grafana Visualization and choose "edit".
4. At the right sidebar menu, go to the "Data Links" section. Click the "+ Add link" button to add a new one:
   1. **Title**: Embrace.

>    2. **URL**: `https://dash-api-us1.embrace.io/external/data_destination/grafana_cloud?labels=${__field.labels}&from=${__from}&to=${__to}`.

Reviewer 2: We really shouldn't be telling people to use dahs-api-us1 hostname, as we were planning to change it. We should put these endpoints behind api.embrace.io

Author: That endpoint lives in dashboard-api, is there an easy way to do it? We have the commitment to do this by Friday (Glovo request). As far as I know, `api.embrace.io` uses Query Service and authentication/authorization which this endpoint doesn't do. 

Also, we already do this, because we have a lot of public endpoints on /external.

Author: We should do something like api.embrace.io/external/data_destination/grafana_cloud hits the dashboard-api.

Author: I saw this one (https://dash-api.embrace.io/status) on the devops repository, should we use that?

Reviewer 2: Use dash-api.embrace.io in this example for now (it's CNAME to -us1), and we'll figure out the rest later. But any moment now, there will be more than one different dash-api URL dependong on the app.

I don't think anyone should have linked to dash-api-us1 because that didn't exist until a few weeks ago.

Author: yes, maybe not dash-api-us1 but for sure `dash-api.embrace.io`
